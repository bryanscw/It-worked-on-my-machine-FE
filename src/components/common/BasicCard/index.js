import React from 'react';
import { Link } from 'react-router-dom';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faEdit, faTrash } from '@fortawesome/free-solid-svg-icons';

export default function BasicCard(props) {
    const {
        classes,
        editable,
        details,
        details: {
            title,
            description,
        },
        badge,
        link,
        handleUpdate,
        handleDelete,
    } = props;

    return (
        <div className={`card${classes ? ` ${classes}` : ""}`}>
            <div className="card-body">
                <h3 className="card-title">
                    <Link to={link}>{title}</Link> {badge}
                </h3>
                <p className="card-text">{description}</p>
                {
                    editable &&
                        <div>
                            <button className="ml-auto btn btn-success mr-2" onClick={() => handleUpdate(details)}>
                                <FontAwesomeIcon icon={faEdit} />
                            </button>
                            <button className="ml-auto btn btn-danger" onClick={() => handleDelete(details)}>
                                <FontAwesomeIcon icon={faTrash} />
                            </button>
                        </div>
                }
            </div>
        </div>
    )
}
