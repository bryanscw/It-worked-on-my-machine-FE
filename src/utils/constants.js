export const API_URL = 'http://localhost:3007';

export const USER_TYPES = {
    STUDENT: 'STUDENT',
    TEACHER: 'TEACHER',
    ADMIN: 'ADMIN',
};

export const CREATE = 'CREATE';
export const UPDATE = 'UPDATE';
export const DELETE = 'DELETE';
export const EMPTY = 'EMPTY';