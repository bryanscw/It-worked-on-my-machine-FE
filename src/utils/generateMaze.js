import generator from 'generate-maze';

export const EMPTY = 0;
export const WALL = 1;
export const GOAL = 2;

function shuffle(array) {
    var currentIndex = array.length, temporaryValue, randomIndex;
  
    // While there remain elements to shuffle...
    while (0 !== currentIndex) {
  
      // Pick a remaining element...
      randomIndex = Math.floor(Math.random() * currentIndex);
      currentIndex -= 1;
  
      // And swap it with the current element.
      temporaryValue = array[currentIndex];
      array[currentIndex] = array[randomIndex];
      array[randomIndex] = temporaryValue;
    }
  
    return array;
}

// Generates a maze of size 29 by 15
export default function generateMaze(numQuestions) {
    const WALL = 1;
    const EMPTY = 0;

    let originalMaze = generator(15, 8);
    let mazeWithWalls = [];

    for (let i = 0; i < originalMaze.length; i++) {
        let originalRow = originalMaze[i];
        let rowWithWalls = [];
        let rowBeneath = [];

        for (let j = 0; j < originalRow.length; j++) {
            rowWithWalls.push(EMPTY);
            rowWithWalls.push(originalRow[j].right ? WALL : EMPTY);
        }

        rowWithWalls.pop()

        for (let j = 0; j < originalRow.length; j++) {
            rowBeneath.push(originalRow[j].bottom ? WALL : EMPTY);
            rowBeneath.push(WALL);
        }

        rowBeneath.pop()

        mazeWithWalls.push(rowWithWalls);
        mazeWithWalls.push(rowBeneath);
    }
    
    mazeWithWalls.pop();

    let lastRow = mazeWithWalls[mazeWithWalls.length - 1];
    mazeWithWalls[mazeWithWalls.length - 1][lastRow.length - 1] = GOAL;

    let freeCoordinates = [];

    for (let i = 0; i < mazeWithWalls.length; i++) {
        let row = mazeWithWalls[i];

        for (let j = 0; j < row.length; j++) {
            if (row[j] === EMPTY && !(i === 0 && j === 0))
                freeCoordinates.push([i, j]);
        }
    }

    freeCoordinates = shuffle(freeCoordinates);
    const questionCoordinates = freeCoordinates.splice(0, numQuestions);
    let mapDescriptor = JSON.stringify(mazeWithWalls);

    return {
        mapDescriptor,
        questionCoordinates,
    };
}