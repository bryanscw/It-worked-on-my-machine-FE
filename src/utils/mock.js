import React from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';

const authReducer = {
    user: {
        email: "test@test.com"
    },
    access_token: "123",
}

export const withRedux = (component, initialState = {}) => {
    const store = createStore(
        (state = {}, action) => state,
        {
            ...initialState,
            authReducer
        },
        applyMiddleware(thunk),
    )

    return <Provider store={store}>{component}</Provider>;
}

export const withRouter = (component, route = '/') => {
    const history = createMemoryHistory({ initialEntries: [route] });
    return <Router history={history}>{component}</Router>;
}

export const withReduxRouter = (component, initialState = {}, route = '/') => {
    const store = createStore(
        (state = {}, action) => state,
        {
            ...initialState,
            authReducer
        },
        applyMiddleware(thunk),
    )
    const history = createMemoryHistory({ initialEntries: [route] });

    return (
        <Provider store={store}>
            <Router history={history}>{component}</Router>
        </Provider>
    );
}