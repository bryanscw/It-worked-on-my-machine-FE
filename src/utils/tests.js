import React from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';

import rootReducer from '../redux/rootReducer';

export const renderWithRedux = (component, initialState = {}) => {
    const store = createStore(
        rootReducer,
        initialState,
        applyMiddleware(thunk),
    )
    return <Provider store={store}>{component}</Provider>;
}

export const renderWithRouter = (component, route = '/') => {
    const history = createMemoryHistory({ initialEntries: [route] });
    return <Router history={history}>{component}</Router>
}

export const renderWithReduxRouter = (component, initialState = {}, route = '/') => {
    return renderWithRedux(renderWithRouter(component, route), initialState);
}