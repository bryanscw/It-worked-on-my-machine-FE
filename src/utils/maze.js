import generator from 'generate-maze';

import { TILE_MAPPING, MAP_HEIGHT, MAP_WIDTH } from './constants';

export const EMPTY = 0;
export const WALL = 1;
export const GOAL = 2;
export const QUESTION = 3;

function shuffle(array) {
    var currentIndex = array.length, temporaryValue, randomIndex;
  
    // While there remain elements to shuffle...
    while (0 !== currentIndex) {
  
      // Pick a remaining element...
      randomIndex = Math.floor(Math.random() * currentIndex);
      currentIndex -= 1;
  
      // And swap it with the current element.
      temporaryValue = array[currentIndex];
      array[currentIndex] = array[randomIndex];
      array[randomIndex] = temporaryValue;
    }
  
    return array;
}

// Generates a maze of size 29 by 15
export function generateMaze(numQuestions) {
    let originalMaze = generator((MAP_WIDTH - 1) / 2, (MAP_HEIGHT - 1) / 2);
    let mazeWithWalls = [];

    for (let i = 0; i < originalMaze.length; i++) {
        let originalRow = originalMaze[i];
        let rowWithWalls = [];
        let rowBeneath = [];

        for (let j = 0; j < originalRow.length; j++) {
            rowWithWalls.push(EMPTY);
            rowWithWalls.push(originalRow[j].right ? WALL : EMPTY);
        }

        rowWithWalls.pop()

        for (let j = 0; j < originalRow.length; j++) {
            rowBeneath.push(originalRow[j].bottom ? WALL : EMPTY);
            rowBeneath.push(WALL);
        }

        rowBeneath.pop()

        mazeWithWalls.push(rowWithWalls);
        mazeWithWalls.push(rowBeneath);
    }
    
    mazeWithWalls.pop();

    // Add goal
    let lastRow = mazeWithWalls[mazeWithWalls.length - 1];
    mazeWithWalls[mazeWithWalls.length - 1][lastRow.length - 1] = GOAL;

    // Add surrounding walls
    let mazeWithSurroundings = mazeWithWalls.map(row => [WALL, ...row, WALL]);
    console.log(mazeWithSurroundings);

    let wallRow = [];
    for (let i = 0; i < MAP_WIDTH; i++)
        wallRow.push(WALL);

    mazeWithSurroundings = [wallRow, ...mazeWithSurroundings, wallRow];

    // Find free coordinates
    let freeCoordinates = [];

    for (let i = 0; i < mazeWithSurroundings.length; i++) {
        let row = mazeWithSurroundings[i];

        for (let j = 0; j < row.length; j++) {
            if (row[j] === EMPTY && !(i === 0 && j === 0))
                freeCoordinates.push([j, i]);
        }
    }

    // Find question coordinats
    freeCoordinates = shuffle(freeCoordinates);
    const questionCoordinates = freeCoordinates.splice(0, numQuestions);

    let mapDescriptor = JSON.stringify(mazeWithSurroundings);

    return {
        mapDescriptor,
        questionCoordinates,
    };
}

export function generateTiles(mapDescriptor, questions) {
    let maze = JSON.parse(mapDescriptor);
    for (let i = 0; i < questions.length; i++) {
        const {
            x,
            y,
        } = questions[i].coordinates;
        maze[y][x] = QUESTION;
    }

    let tiles = [];
    for (let i = 0; i < maze.length; i++) {
        let row = maze[i];
        let tileRow = [];

        for (let j = 0; j < row.length; j++) {
            let tile = null;

            switch (row[j]) {
                case EMPTY:
                    tile = TILE_MAPPING.grass;
                    break;

                case WALL:
                    tile = Math.random() <= 0.75 ? TILE_MAPPING.tree: TILE_MAPPING.rock;
                    break;

                case GOAL:
                    tile = TILE_MAPPING.house;
                    break;

                case QUESTION:
                    tile = Math.floor(Math.random() * 9) + 7;
                    break;
                
                default:
                    tile = TILE_MAPPING.grass;
                    break;
            }

            tileRow.push(tile);
        }

        tiles.push(tileRow);
    }

    return tiles;
}