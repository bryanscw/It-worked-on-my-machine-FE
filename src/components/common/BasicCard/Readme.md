
```js
import 'bootstrap/dist/css/bootstrap.css';
import 'bootstrap/dist/js/bootstrap.js';

import { METHODS } from '../../../redux/ducks/apiHelper.js';
import { withReduxRouter } from '../../../utils/mock.js';
const details = {
        title: 'Basic Addition',
        description: 'This is an interesting video',
    };
withReduxRouter(<BasicCard editable={true} details={details} link="https://www.youtube.com/embed/AuX7nPBqDts" handleUpdate={()=>{}} handleDelete={()=>{}}/>)
```