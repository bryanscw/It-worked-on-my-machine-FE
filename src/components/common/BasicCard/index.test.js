import React from 'react'
import { render, cleanup } from '@testing-library/react'

import BasicCard from './index.js';
import { renderWithRouter } from '../../../utils/tests';

afterEach(cleanup);

it('should take a snapshot', async () => {
    const details = {
        title: 'Basic Addition',
        description: 'This is an interesting video',
    };

    const { asFragment } = renderWithRouter(<BasicCard editable={false} details={details} link="" />);
    expect(asFragment()).toMatchSnapshot();
})