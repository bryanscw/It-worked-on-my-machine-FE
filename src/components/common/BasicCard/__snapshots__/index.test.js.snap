// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <div
      class="card-body"
    >
      <h3
        class="card-title"
      >
        <a
          href="/"
        >
          Basic Addition
        </a>
         
      </h3>
      <p
        class="card-text"
      >
        This is an interesting video
      </p>
    </div>
  </div>
</DocumentFragment>
`;
